clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumer.java

```diff
@@ -209,90 +218,140 @@ private AbstractAtLeastOnceKafkaConsumer(final KafkaConsumer<String, Buffer> kaf
     }
 
     /**
-     * Closes the underlying Kafka consumer.
+     * Closes the underlying Kafka consumer and tries to commit the current offsets before.
      * <p>
      * This does not invoke the close handler.
      *
-     * @return A future that will complete when the Kafka consumer is closed.
+     * @return A future that will complete when the Kafka consumer is closed. If the offsets have been committed
+     *         successfully, the future will succeed, otherwise, it will fail.
      */
     @Override
     public Future<Void> stop() {
-        final Promise<Void> promise = Promise.promise();
-        kafkaConsumer.close(promise);
-        return promise.future();
+        return tryCommitAndClose();
     }
 
     private void handleBatch(final KafkaConsumerRecords<String, Buffer> records) {
-
-        LOG.debug("Polled {} records", records.size());
-        CompositeFuture.all(processBatch(records))
-                .compose(ok -> commit(true))
-                .compose(ok -> poll())
-                .onSuccess(this::handleBatch);
-
+        try {
+            LOG.debug("Polled {} records", records.size());
+
+            for (int i = 0; i < records.size(); i++) {
+                if (stopped) {
+                    return;
+                }
+
+                final KafkaConsumerRecord<String, Buffer> record = records.recordAt(i);
+                final T message = createMessage(record);
+                try {
+                    messageHandler.handle(message);
+                    addToCurrentOffsets(record);
+                } catch (final RuntimeException messageHandlingError) {
+                    LOG.debug("Message handler failed", messageHandlingError);
+                    // will commit the offset of the failed record and then resume polling (will include failed record)
+                }
```

Comment:
why are we processing the remaining records here? we want to commit the offsets up to the first record that has caused the exception, don't we?

Reply:
Uh, that's bad. You are totally right. I missed adding a `break` statement :worried: